src/librustc_lint/builtin.rs

```diff
@@ -2087,6 +2087,82 @@ impl LintPass for InvalidNoMangleItems {
     }
 }
 
+#[derive(Clone, Copy)]
+pub struct MutableTransmutes;
+
+declare_lint! {
+    MUTABLE_TRANSMUTES,
+    Deny,
+    "transmuting from &T to &mut T is undefined behavior"
+}
+
+impl LintPass for MutableTransmutes {
+    fn get_lints(&self) -> LintArray {
+        lint_array!(MUTABLE_TRANSMUTES)
+    }
+
+    fn check_expr(&mut self, cx: &Context, expr: &ast::Expr) {
+        use syntax::ast::DefId;
+        use syntax::abi::RustIntrinsic;
+        let msg = "transmuting from immutable to mutable is undefined behavior";
+        match get_transmute_from_to(cx, expr) {
+            Some((&ty::ty_rptr(_, from_mt), &ty::ty_rptr(_, to_mt))) => {
+                if to_mt.mutbl == ast::Mutability::MutMutable
+                    && from_mt.mutbl == ast::Mutability::MutImmutable {
+                    cx.span_lint(MUTABLE_TRANSMUTES, expr.span, msg);
+                }
+            }
+            _ => ()
+        }
+
+        fn get_transmute_from_to<'a, 'tcx>(cx: &Context<'a, 'tcx>, expr: &ast::Expr)
+            -> Option<(&'tcx ty::sty<'tcx>, &'tcx ty::sty<'tcx>)> {
+            if let ast::ExprPath(..) = expr.node {
+                if let DefFn(did, _) = ty::resolve_expr(cx.tcx, expr) {
+                    if def_id_is_transmute(cx, did) {
+                        let typ = ty::node_id_to_type(cx.tcx, expr.id);
+                        match typ.sty {
+                            ty::ty_bare_fn(_, ref bare_fn) if bare_fn.abi == RustIntrinsic => {
+                                if let ty::FnConverging(to) = bare_fn.sig.0.output {
+                                    let from = bare_fn.sig.0.inputs[0];
+                                    return Some((&from.sty, &to.sty));
+                                }
+                            },
+                            _ => ()
+                        }
+                    }
+                }
+            }
+            None
+        }
+
+        fn def_id_is_transmute(cx: &Context, def_id: DefId) -> bool {
+            use syntax::ast_map::NodeForeignItem;
+            let intrinsic = match ty::lookup_item_type(cx.tcx, def_id).ty.sty {
+                ty::ty_bare_fn(_, ref bfty) => bfty.abi == RustIntrinsic,
+                _ => return false
+            };
+            if def_id.krate == ast::LOCAL_CRATE {
+                match cx.tcx.map.get(def_id.node) {
+                    NodeForeignItem(ref item) if intrinsic => {
+                        token::get_ident(item.ident) ==
```

Comment:
Why compare the strings? It seems `token::get_ident()` doesn't return an `ast::Ident` (the fn name is misleading), but an `InternedString`. Comparing an interned string should be strictly faster than pulling out the original Strings and then doing a memcmp, right?